Dev 4.x nevm

sidhujag opened the pull request.

Commit:
NEVM initial WIP

setup ZMQ comm. mechanism
store nevm data in block/index
prune data from index in chainlock
support transport of the nevm data over network
validate calls backs of nevm data through zmq layer
zmq plumbing for nevm calls (miner, connect/disconnect)

Commit:
fix trailing whitespace

Commit:
fix CI

Commit:
fix CI

Commit:
clean up cevmblock

Commit:
make DecodeHexBlockHeader/DecodeHexBlk more resilient

check if its valid, fallback to ser_network

Commit:
add disabled test back

Commit:
fix pubsub with zmq_connect for subscriber

Commit:
fix test

Commit:
Update gitian-osx.yml

Commit:
Update 00_setup_env_mac.sh

Commit:
multiple fixes on pruning evm data + enforcement

-cleanup blocktreedb on load/save for when blocks are older than 2*nMaxTipAge
-check for non-empty nevm data in AcceptBlockHeader if not IBD
-flag enforcement of nevm checks through nNEVMEnforce chainparams flag that can be adjusted on init

Commit:
initial zmq nevm test

Commit:
Update init.cpp

Commit:
Update validation.cpp

Commit:
add some comments

Commit:
fix zmq interface test

Commit:
fix data race

Commit:
strict ConnectNEVMCommitment duplicate check

ensure evm blocks are all checked unless IBD, ensure duplicate evm hashes are discarded upront

Commit:
fix zmq test context issue + CI lint

context issue saw us dbl closing socket, removed explicit closing and assume exception closes out sockts

Commit:
fix CI

Commit:
Update interface_zmq_nevm.py

Commit:
add block mappings to zmq test

remove processNewBlock logic from connectblock - only depend on ibd nothing else

Commit:
nevm enforce through zmq connection

+ enforce masternode to define those connect/disconnect zmq messages

Commit:
Update init.cpp

Commit:
add second node on test to have connect/disconnect zmq handlers

Commit:
Update interface_zmq_nevm.py

Commit:
fix CI

Commit:
add buffer of time on p2p evm data

pruning happens after 2*maxtip, consensus enforces that within IBD (1*maxtip) the evm data exists, and that > 3*maxtip that it doesn't exist. Therefor we need to propogate if header is within 2*maxtip (if its set), so chances are next to none that 1*maxtip / 3*maxtip rule will trigger.

also move the evm data checks in ContextualCheckBlockHeader which is checked in the miner prior to creating and sending block.

Commit:
fix CI

Commit:
Update test_runner.py

Commit:
create block and disk index for nevm block data

create indexes for block data so we can use it as needed for various things like lookups and recursing backwards to prune older indexes on chainlocks etc

Commit:
internalize reading nevm block index + getblockheader

fix intermittent compactblock errors
add strict check for block data empt is waiting for response
support reindex/onload validation skips where block data is empty by using fLoaded for bWaitForResponse flag

Commit:
Update chain.cpp

Commit:
Update validation.cpp

Commit:
add waitforresponse to nevm connect/disconnect messages

we need to know if sys node is waiting for a response otherwise it will send messages non-blocking and on subsequent calls end up receiving the wrong one when it is blocking.

It might be also useful to know when geth node can sync up with other nodes (when responses are expected it means IBD is done)

add reorg test to zmq nevm test

Commit:
re-org with artifical delay test

Commit:
Update interface_zmq_nevm.py

fanquake committed:
Merge #22339: doc: Document minimum required libc++ version

Commit:
Merge bitcoin/bitcoin#22146: Reject invalid coin height and output index when loading assumeutxo

fa9ebedec3f982bb5bb459ea33d74c94d9b5cec4 Reject invalid coin height and output index when loading assumeutxo (MarcoFalke)

Pull request description:

  It should be impossible to have a coin at a height higher than the height of the snapshot block, so reject those early to avoid integer wraparounds and hash collisions later on.

  Same for the outpoint index.

  Both issues were found by fuzzing:

  * The height issue by OSS-Fuzz: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=34793
  * The outpoint issue by my fuzz server: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=34793#c2

ACKs for top commit:
  practicalswift:
    cr ACK fa9ebedec3f982bb5bb459ea33d74c94d9b5cec4: patch looks correct
  jamesob:
    crACK https://github.com/bitcoin/bitcoin/pull/22146/commits/fa9ebedec3f982bb5bb459ea33d74c94d9b5cec4
  theStack:
    Code review ACK fa9ebedec3f982bb5bb459ea33d74c94d9b5cec4
  benthecarman:
    crACK fa9ebedec3f982bb5bb459ea33d74c94d9b5cec4

Tree-SHA512: dae7caee4b3862b23ebdf2acb7edec4baf75b0dbf1409b370b1a73aa6b632b317ebfac596dcbaf4edfb1301b513f45465ea75328962460f35e2af0d7e547c9ac

Commit:
CNEVMBlockDisconnect which is only a single hash

disconnect only needs sys hash for geth node to operate and work on disconnecting, it can internally lookup the geth block hash through reverse map

Commit:
Update zmqpublishnotifier.cpp

Commit:
cleanup nevm zmq to reuse sockets + geth startup logic

Commit:
skip over nevm validation when reverifying blocks on startup, also start geth prior to zmq

enforce zmq connects to all sockets on init if in nevm enforcement mode (fNEVMConnection)

Commit:
rework zmq notifications

1) use req/rep model
2) add nevmcomms to trigger startup/shutdown events
3) remove waitforresponse in connect/disconnects since we use req/rep model

Commit:
fix CNEVMBlockIndex

1) we did not have the data actually being indexed which may cause problems on syncing with other nodes since data would have been pruned within maxtips*2 blocks from tip
2) save nTime in CNEVMBlockIndex which allowed serialization to work properly according to the tip*2 test
3) store hash not pointer to hash in CNEVMBlockIndex since we don't have full header to recreate hash

Commit:
zmq fixes

1) don't fail and close socket if nevm calls fail for whatever reason, they are allowed to fail
2) set seperate socket timeouts depending on nevm message, 60 seconds on recv connect, 30 seconds on disconnect and 5 seconds on comms

Commit:
pause to let geth catch up to connect for 5 seconds

Commit:
add nevm parent block hash

to support contigious checks around mappings

Commit:
SanitizeGethCmdLine fixes

remove SanitizeGethCmdLine overrides

Commit:
add virtual dtor's for scheme's in new BLS impl

Commit:
multiple NEVM related fixes

1) fix precision parsing (removed bridge id)
2) fix syscoingetspvproof
3) fix placement of ConnectNEVMCommitment (outside of tx loop)
4) update vchTokenFreezeMethod to new correct hash

Commit:
add nodiscover by default since we use v5disc, it disables v4

Commit:
use testnet descriptor for regtest

Commit:
update guix naming and add missing config params to .configure-custom.sh

Commit:
remove assert which is never false

Commit:
update version

Commit:
set block version to note nevm info

Commit:
rm files

Commit:
sync testnet

Commit:
fix superblock cycle to account for 2.5x

Commit:
Update schemes.hpp

Commit:
Revert "add virtual dtor's for scheme's in new BLS impl"

This reverts commit 49d545510ffe4bc452bc1b1f97a8c31bf4210ba5.

Commit:
Revert "Revert "add virtual dtor's for scheme's in new BLS impl""

This reverts commit 03e381dc7f3a91c1186cefa2b4b81d9434dd829f.

Commit:
Update schemes.hpp

Commit:
parameterize halving fn and powtargetspacing/interval functions

Commit:
logging

Commit:
validation fix when MN Floor is reached

Commit:
fix lock order running deterministicmn test

Commit:
fix minmn check

Commit:
Revert "logging"

This reverts commit 8077838f61951fd67a0072d687b24014e7105ffb.

Commit:
use v4 discovery for now

Commit:
Update chainparams.cpp

Commit:
Revert "use v4 discovery for now"

This reverts commit 92b17bf8da4ddfba983c785002a84e44edeb9e77.

Commit:
rm nodiscover and v5disc default

Commit:
only set evm block data if not in IBD

if we push data outside of IBD then geth starts syncing blocks before mappings are set

Commit:
persistent nodekey file

we should preserve nodekey file in geth during reindex's as services shouldn't have to expose new nodekey's externally on reindex's

Commit:
refer to master descriptor URl always

Commit:
create fake testnet

Commit:
set params for fake testnet

Commit:
update testnet params for test

Commit:
check for testnet sig vs mainnet in descriptor spec

Commit:
shutdown on block if exitwhensynced is defined for geth params

Commit:
better killprocess

send one signal and wait instead of repeated calls

Commit:
on start check if geth already running and shutdown prior to restart

Commit:
on startup on osx/linux check if straggling sysgeth process already running, kill it

Commit:
always just kill any sysgeth already running on start

Commit:
shut down on any "connected" errors if exitwhensynced is set

Commit:
avoid sigterm/sigkill if process does not exist

Commit:
do not mark block invalid if exitwhensynced error happens

Commit:
check for process exit at != 0

Commit:
rmv tmp files

Commit:
wrap mn payment pct

Commit:
startup param for geth

willyko commented on src/masternode/masternodepayments.cpp:
```diff
@@ -215,9 +216,17 @@ bool CMasternodePayments::GetMasternodeTxOuts(CChain& activeChain, int nBlockHei
     CAmount nSubsidy = nBlockReward*0.75;
+    
+    // ensure that if subsidy is less than min amount then stick to min for long term MN full node/chainlock incentive
+    const CAmount &nMinMN = 527500000;
```
better to put this in chainparams?

sidhujag replied:
done

Commit:
Update params.h

Commit:
fix nonfatal exception on moneyrange for burn asset to syscoin txs

these end up giving negative fee so we need to account for the first output in a burn asset to syscoin tx so that the fee is correctly calculated and avoid the nonfatal exception.

Commit:
added getnevmblockchaininfo RPC

Commit:
rm files

Commit:
add mn min subsidy sats to chainparams

Commit:
add status to getnevmblockchaininfo

send comms message and check for ack in status

Commit:
recycle zmq services upon geth start

require subs and pubs to recycle to geth node will get new sockets

Commit:
rm tmp files

Commit:
reset last block on syscoinstartgeth

reconsider prev block if its invalid, assuming that is the reason for the failure (because zmq is not enabled and geth is offline)

Commit:
rm

Commit:
Merge branch 'dev-4.x' into dev-4.x-nevm
